libraries/adaptive-expressions/src/builtinFunctions/formatNumber.ts

```diff
@@ -21,31 +23,944 @@ export class FormatNumber extends ExpressionEvaluator {
     }
 
     private static evaluator(): EvaluateExpressionDelegate {
-        return FunctionUtils.applyWithError(
-            (args: any[]): any => {
+        return FunctionUtils.applyWithOptionsAndError(
+            (args: any[], options: Options): any => {
                 let value: any = null;
                 let error: string;
                 let number = args[0];
                 let precision = args[1];
-                let locale = args.length > 2 ? args[2] : 'en-us';
+                let locale = options.locale ? options.locale : 'en-us';
+                locale = FunctionUtils.determineLocale(args, locale, 3);
                 if (typeof number !== 'number') {
                     error = `formatNumber first argument ${number} must be a number`;
                 } else if (typeof precision !== 'number') {
                     error = `formatNumber second argument ${precision} must be a number`;
                 } else if (locale && typeof locale !== 'string') {
                     error = `formatNubmer third argument ${locale} is not a valid locale`;
                 } else {
-                    // NOTE: Nodes toLocaleString and Intl do not work to localize unless a special version of node is used.
-                    // TODO: In R10 we should try another package.  Numeral and d3-format have the basics, but no locale specific.  
-                    // Numbro has locales, but is optimized for the browser.
-                    value = number.toLocaleString(locale, { minimumFractionDigits: precision, maximumFractionDigits: precision });
+                    const fixedNotation = `,.${precision}f`;
+                    const roundedNumber = this.roundToPrecision(number, precision);
+                    const formatLocale = FormatNumber.localeSets[locale];
+                    if (formatLocale !== undefined) {
+                        value = d3.formatLocale(formatLocale).format(fixedNotation)(roundedNumber);
+                    } else {
+                        value = d3.format(fixedNotation)(roundedNumber);
+                    }
                 }
 
                 return { value, error };
             });
     }
 
+    private static roundToPrecision = (num: number, digits: number): number => Math.round(num * Math.pow(10, digits)) / Math.pow(10, digits);
+
     private static validator(expr: Expression): void {
         FunctionUtils.validateOrder(expr, [ReturnType.String], ReturnType.Number, ReturnType.Number);
     }
+
+    private static localeSets = {
```

Comment:
These localeSets are odd to me.  Is there not a library we can use which contains this functionality?  

If we do use this method, this list should be in its own file.

Reply:
In the previous implementation of formatNumber, there is a comment there.
```
Nodes toLocaleString and Intl do not work to localize unless a special version of node is used.
Numeral and d3-format have the basics, but no locale specific.
Numbro has locales, but is optimized for the browser.  
```
The reason I choose d3-format is they provided an API to support formatLocale. However in d3-format, they stored the locale info in external files, see https://github.com/d3/d3-format
To ensure our Adaptive-Expressions can work on both server and browser, I transformed these files into a dictionary. 

The reason I placed these localeSet as a private property here is these locale info is only applied here. And it is not appropriate to make a file under builtinFunctions folder. This folder should only contain file of built-in functions.